File: clp/minimal_loc/b_trees.pl
inInclusiveRange(Start, End, Start) :-
    Start =< End.
inInclusiveRange(Start, End, Result) :-
    Start < End,
    NewStart is Start + 1,
    inInclusiveRange(NewStart, End, Result).
inInclusiveRangeList(Start, End, []) :-
    Start > End.
inInclusiveRangeList(Start, End, [Start|Rest]) :-
    Start =< End,
    NewStart is Start + 1,
    inInclusiveRangeList(NewStart, End, Rest).
yolo_UNSAFE_minChildSize(H, Output) :-
    HMinusOne is H - 1,
    powInt(2, HMinusOne, PowResult),
    Output is 2 * PowResult - 1.
yolo_UNSAFE_maxChildSize(H, Output) :-
    powInt(4, H, Temp),
    Output is Temp - 1.
getAdditions(0, Amount, _, _) :-
    Amount > 0,
    !,
    fail.
getAdditions(0, _, _, []) :- !.
getAdditions(1, Amount, Max, _) :-
    Amount > Max,
    !,
    fail.
getAdditions(Size, Amount, Max, [Added|Rest]) :-
    RangeEnd is min(Amount, Max),
    SizeMinusOne is Size - 1,
    inInclusiveRange(0, RangeEnd, Added),
    AmountMinusAdded is Amount - Added,
    getAdditions(SizeMinusOne, AmountMinusAdded, Max, Rest).
sublist([], []).
sublist([H|T], [H|Rest]) :-
    sublist(T, Rest).
sublist([_|T], Rest) :-
    sublist(T, Rest).
sublistOfSize(Input, Size, Output) :-
    length(Output, Size),
    sublist(Input, Output).
yolo_UNSAFE_rootBTreeHeight(Size, Height) :-
    SizePlusOne is Size + 1,
    logInt(SizePlusOne, LogResult),
    LogResultPlusOne is LogResult + 1,
    inInclusiveRange(1, LogResultPlusOne, Height).
makeRootBTree(Size, Result) :-
    yolo_UNSAFE_rootBTreeHeight(Size, Height),
    makeRootBTree(Size, 1, Size, Height, Result).
makeRootBTree(Size, KeyRangeStart, KeyRangeEnd, 1, b_tree(X, [])) :-
    Size < 4,
    !,
    inInclusiveRangeList(KeyRangeStart, KeyRangeEnd, KeyRangeList),
    sublistOfSize(KeyRangeList, Size, X).
makeRootBTree(Size, KeyRangeStart, KeyRangeEnd, H, Result) :-
    H > 1,
    !,
    inInclusiveRange(2, 4, X),
    makeBTree(X, Size, KeyRangeStart, KeyRangeEnd, H, Result).
makeNonRootBTree(Size, KeyRangeStart, KeyRangeEnd, 1, b_tree(X, [])) :-
    Size < 4,
    Size >= 1,
    !,
    inInclusiveRangeList(KeyRangeStart, KeyRangeEnd, KeyRangeList),
    sublistOfSize(KeyRangeList, Size, X).
makeNonRootBTree(Size, KeyRangeStart, KeyRangeEnd, H, Result) :-
    H > 1,
    Size > 0,
    !,
    inInclusiveRange(2, 4, X),
    makeBTree(X, Size, KeyRangeStart, KeyRangeEnd, H, Result).
map([], _, []).
map([A|As], Relation, [B|Bs]) :-
    call(Relation, A, B),
    map(As, Relation, Bs).
zip([], [], []).
zip([A|As], [B|Bs], [pair(A, B)|Rest]) :-
    zip(As, Bs, Rest).
scanLeft([], B, _, [B]).
scanLeft([A|As], B, Relation, [B|Bs]) :-
    call(Relation, B, A, NewB),
    scanLeft(As, NewB, Relation, Bs).
tail([_|Rest], Rest).
init([_], []).
init([A1, A2|As], [A1|Rest]) :-
    init([A2|As], Rest).
keysHelper(ChildSizes, AddKeys, KeyRangeStart, Keys) :-
    zip(ChildSizes, AddKeys, ZipResult),
    InitialScan is KeyRangeStart - 1,
    scanLeft(ZipResult,
             InitialScan,
             lambda([SoFar,
                     pair(ChildSize, Add),
                     LambdaResult],
                    LambdaResult is SoFar + ChildSize + Add + 1),
             ScanResult),
    tail(ScanResult, TailResult),
    init(TailResult, Keys).
childRangesHelper(KeyRangeStart, Keys, KeyRangeEnd, ChildRanges) :-
    StartMinusOne is KeyRangeStart - 1,
    EndPlusOne is KeyRangeEnd + 1,
    append(Keys, [EndPlusOne], ZipRight),
    zip([StartMinusOne|Keys], ZipRight, ZipResult),
    map(ZipResult,
        lambda([pair(A, B),
                pair(InnerStart, InnerEnd)],
               (InnerStart is A + 1,
                InnerEnd is B - 1)),
        ChildRanges).
childrenHelper(ChildSizes, ChildRanges, HMinusOne, Children) :-
    zip(ChildSizes, ChildRanges, ZipResult),
    map(ZipResult,
        lambda([pair(Size, pair(RangeStart, RangeEnd)),
                Child],
               makeNonRootBTree(Size, RangeStart, RangeEnd, HMinusOne, Child)),
        Children).
makeBTree(NChildren, Size, KeyRangeStart, KeyRangeEnd, H, b_tree(Keys, Children)) :-
    HMinusOne is H - 1,
    yolo_UNSAFE_minChildSize(HMinusOne, MinChildSizeBelow),
    yolo_UNSAFE_maxChildSize(HMinusOne, MaxChildSizeBelow),
    NKeys is NChildren - 1,
    RestOfNodes is Size - NKeys - NChildren * MinChildSizeBelow,
    RestOfNodes >= 0,
    KeyRangeSize is KeyRangeEnd - KeyRangeStart + 1,
    AddKeysAmount is KeyRangeSize - Size,
    AddListMax is MaxChildSizeBelow - MinChildSizeBelow,
    getAdditions(NChildren, RestOfNodes, AddListMax, AddList),
    map(AddList, lambda([Current, CurrentAdded], CurrentAdded is Current + MinChildSizeBelow), ChildSizes),
    getAdditions(NChildren, AddKeysAmount, KeyRangeSize, AddKeys),
    keysHelper(ChildSizes, AddKeys, KeyRangeStart, Keys),
    childRangesHelper(KeyRangeStart, Keys, KeyRangeEnd, ChildRanges),
    childrenHelper(ChildSizes, ChildRanges, HMinusOne, Children).
clausedef(b_tree_main, [], [int, b_tree]).
b_tree_main(Size, Tree) :-
    makeRootBTree(Size, Tree).

File: src/bench/test/scala/scife/enumeration/benchmarks/iterators/b_trees.scala
package scife.enumeration.benchmarks.iterators

import scife.enumeration.memoization.MemoizationScope
import scife.util.structures.BTree._

import Iterators._

class BTreesBenchmark(key: String, makeCache: () => IteratorCache[Any, Tree]) extends BenchmarkAdapter[Any, Tree](key, makeCache) {
  def makeIterator(size: Int, cache: IteratorCache[Any, Tree]): Iterator[Tree] = {
    BTrees.makeRootBTree(size)(cache)
  }
  def makeMeasureIterator(size: Int, cache: IteratorCache[Any, Tree]): Iterator[Tree] = {
    makeIterator(size, cache)
  }
  def makeWarmUpIterator(size: Int, cache: IteratorCache[Any, Tree]): Iterator[Tree] = {
    makeIterator(size, cache)
  }
}

object BTrees {
  val t = 2

  def minChildSize(h: Int): Int = 2 * math.pow(t, h - 1).toInt - 1
  def maxChildSize(h: Int): Int = math.pow(2 * t, h).toInt - 1

  def getAdditions(size: Int, amount: Int, max: Int): Iterator[List[Int]] = {
    if (size == 0 && amount > 0) EmptyIterator
    else if (size == 0) Nil
    else if (size == 1 && amount > max) EmptyIterator
    else {
      for {
	added <- 0.to(math.min(amount, max)).iterator
	rest <- getAdditions(size - 1, amount - added, max)
      } yield added :: rest
    }
  }

  // sublist([], []).
  // sublist([H|T], [H|Rest]) :-
  //     sublist(T, Rest).
  // sublist([_|T], Rest) :-
  //     sublist(T, Rest).
  def sublists[A](list: List[A]): Iterator[List[A]] = {
    if (list.isEmpty) {
      Nil
    } else {
      sublists(list.tail).flatMap(rest =>
	anyToIterator(list.head :: rest).or(rest))
    }
  }

  // the algorithm in SciFe is incredibly complex for this, and I don't completely
  // understand it.  Here, we do a simple filter with the results of sublists
  def sublistsOfSize[A](list: List[A], size: Int): Iterator[List[A]] = {
    sublists(list).filter(_.size == size)
  }

  // for the initial call
  def makeRootBTree(size: Int)(implicit cache: IteratorCache[Any, Tree]): Iterator[Tree] = {
    for {
      height <- 1.to(math.log(size + 1).toInt + 1).iterator
      result <- makeRootBTree(size, 1.to(size), height)
    } yield result
  }

  def makeRootBTree(size: Int, keyRange: Range, h: Int)(implicit cache: IteratorCache[Any, Tree]): Iterator[Tree] = {
    def uncachedIterator: Iterator[Tree] = {
      if (h == 1 && size < 2 * t) {
	for {
	  x <- sublistsOfSize(keyRange.toList, size)
	} yield Tree(x, Nil)
      } else if (h > 1) {
	for {
	  x <- 2.to(2 * t).iterator
	  result <- makeBTree(x, size, keyRange, h)
	} yield result
      } else EmptyIterator
    }
    cache.tryCache((size, keyRange, h), uncachedIterator)
  }

  def makeNonRootBTree(size: Int, keyRange: Range, h: Int)(implicit cache: IteratorCache[Any, Tree]): Iterator[Tree] = {
    def uncachedIterator: Iterator[Tree] = {
      if (h == 1 && size < 2 * t && size >= t - 1) {
	for {
	  x <- sublistsOfSize(keyRange.toList, size)
	} yield Tree(x, Nil)
      } else if (h > 1 && size > 0) {
	for {
	  x <- t.to(2 * t).iterator
	  result <- makeBTree(x, size, keyRange, h)
	} yield result
      } else EmptyIterator
    }
    cache.tryCache((size, keyRange, h), uncachedIterator)
  }

  def liftMap[A, B](list: List[A], makeIterator: A => Iterator[B]): Iterator[List[B]] = {
    list match {
      case a :: as => {
	for {
	  b <- makeIterator(a)
	  rest <- liftMap(as, makeIterator)
	} yield b :: rest
      }
      case Nil => Nil
    }
  }

  def makeBTree(nChildren: Int, size: Int, keyRange: Range, h: Int)(implicit cache: IteratorCache[Any, Tree]): Iterator[Tree] = {
    def uncachedIterator: Iterator[Tree] = {
      val minChildSizeBelow: Int = minChildSize(h - 1)
      val maxChildSizeBelow: Int = maxChildSize(h - 1)
      val nKeys: Int = nChildren - 1
      val restOfNodes: Int = size - nKeys - nChildren * minChildSizeBelow
      
      if (restOfNodes < 0) {
	EmptyIterator
      } else {
	for {
	  addList <- getAdditions(nChildren, restOfNodes, maxChildSizeBelow - minChildSizeBelow)
	  childSizes: List[Int] = addList.map(_ + minChildSizeBelow)
	  addKeys <- getAdditions(nChildren, keyRange.size - size, keyRange.size)
	  keys: List[Int] = childSizes.zip(addKeys).scanLeft(keyRange.start - 1)(
	    { case (soFar, (childSize, add)) => soFar + childSize + add + 1 }).tail.init
	  childRanges: List[Range] = ((keyRange.start - 1) :: keys).zip(keys :+ (keyRange.end + 1)).map(
	    { case (a, b) => (a + 1).to(b - 1) })
	  children <- liftMap(childSizes.zip(childRanges),
			      (pair: (Int, Range)) => makeNonRootBTree(pair._1, pair._2, h - 1))
	  // children: List[Iterator[Tree]] = childSizes.zip(childRanges).map(
	  //   { case (cs, cr) => makeNonRootBTree(cs, cr, h - 1) })
	} yield Tree(keys, children)
      }
    } // uncachedIterator

    cache.tryCache((nChildren, size, keyRange, h), uncachedIterator)
  }

  def expected(size: Int): List[Tree] = {
    import scife.enumeration.Enum
    import scife.enumeration.dependent.Depend
    import scife.enumeration.benchmarks.BTreeTest

    val dep: Depend[(Int, Range, Int), Tree] =
      (new BTreeTest).constructEnumerator
    for {
      height <- 1.to(math.log(size + 1).toInt + 1).toList
      enum = dep.apply((size, 1.to(size), height))
      ind <- 0.until(enum.size).toList
    } yield enum(ind)
  }

  def walkThrough(size: Int, memo: MemoizationScope): Long = {
    import scife.enumeration.Enum
    import scife.enumeration.dependent.Depend
    import scife.enumeration.benchmarks.BTreeTest

    var numStructures: Long = 0

    val dep: Depend[(Int, Range, Int), Tree] =
      (new BTreeTest).constructEnumerator(memo)

    1.to(math.log(size + 1).toInt + 1).foreach(height => {
      val enum = dep.apply((size, 1.to(size), height))
      enum.foreach(_ => numStructures += 1)
    })
    numStructures
  }
    
  // takes a bound, and makes sure that we agree with SciFe
  def main(args: Array[String]) {
    if (args.length != 1) {
      println("Needs a size bound")
    } else {
      val size = args(0).toInt
      val exList = expected(size).toList
      val gotList = makeRootBTree(size)(new RealCache).toList
      if (exList.size != gotList.size) {
	println("RAW LISTS DIFFERED IN SIZES")
      } else {
	println("NUM ELEMENTS: " + gotList.size)
	// println("ELEMENTS: " + gotList)
      }
      val ex = exList.toSet
      val got = gotList.toSet
      if (ex != got) {
	val missing = ex -- got
	val extra = got -- ex
	println("MISSING: " + missing)
	println("EXTRA: " + extra)
      }
      println("SIZE: " + gotList.size)
      gotList.foreach(println)
    }
  } // main
} // BTrees	
